File: libgomp/testsuite/libgomp.fortran/task-detach-1.f90
! { dg-do run }

! Test chaining of detached tasks, with each task fulfilling the
! completion event of the previous one.

program task_detach_1
  use omp_lib

  integer (kind=omp_event_handle_kind) :: detach_event1, detach_event2
  integer :: x = 0, y = 0, z = 0

  !$omp parallel
    !$omp single
      !$omp task detach (detach_event1)
        x = x + 1
      !$omp end task

      !$omp task detach (detach_event2)
        y = y + 1
	call omp_fulfill_event (detach_event1)
      !$omp end task

      !$omp task
        z = z + 1
	call omp_fulfill_event (detach_event2)
      !$omp end task
    !$omp end single
  !$omp end parallel

  if (x /= 1) stop 1
  if (y /= 1) stop 2
  if (z /= 1) stop 3
end program

File: libgomp/testsuite/libgomp.c-c++-common/task-detach-1.c
/* { dg-do run } */

#include <omp.h>
#include <assert.h>

/* Test chaining of detached tasks, with each task fulfilling the
   completion event of the previous one.  */

int main (void)
{
  omp_event_handle_t detach_event1, detach_event2;
  int x = 0, y = 0, z = 0;

  #pragma omp parallel
    #pragma omp single
    {
      #pragma omp task detach (detach_event1)
	x++;

      #pragma omp task detach (detach_event2)
      {
	y++;
	omp_fulfill_event (detach_event1);
      }

      #pragma omp task
      {
	z++;
	omp_fulfill_event (detach_event2);
      }
    }

  assert (x == 1);
  assert (y == 1);
  assert (z == 1);
}
